### ISSUE TITLE
Help with hosting

### ISSUE THREAD

**Issue by manulera** [manulera/YeastDatabase#7]
Hi @NicolasCARPi ,

Once again asking for help.

I am trying to deploy the site in digital ocean (this is just because I am applying for a grant and I would like to be able to provide a link to the prototype, even though I think I will not keep developing this specific version, it shows the use-case).

I managed to do it in digital ocean with http with docker, and I got it working with [this docker-compose](https://github.com/manulera/YeastDatabase/blob/a5d7d09dd2de4302077faf6d4733e8d44a7c1349/docker-compose.yml), [this docker file](https://github.com/manulera/YeastDatabase/blob/a5d7d09dd2de4302077faf6d4733e8d44a7c1349/Dockerfile) and [this nginx configuration](https://github.com/manulera/YeastDatabase/blob/a5d7d09dd2de4302077faf6d4733e8d44a7c1349/phpdocker/nginx/nginx.conf). It can be accessed [here](http://www.prototype.genestorian.org/).

I am now trying to get the ssl certificates to make it work with https, but I just can't manage. I looked around, and it seemed in a bunch of forums people follow [this blog post](https://pentacent.medium.com/nginx-and-lets-encrypt-with-docker-in-less-than-5-minutes-b4b8a60d3a71), with [this docker-compose file](https://github.com/manulera/YeastDatabase/blob/master/docker-compose.yml)  and using [this bash script](https://raw.githubusercontent.com/wmnnd/nginx-certbot/master/init-letsencrypt.sh) in which you have to set a few things. This does not work for me and I just can't figure it out. The error that I get is below. Let me know if anything rings a bell, or if you can point me to a blog post or something to figure it out.

Thanks!

```
Requesting a certificate for prototype.genestorian.org and www.prototype.genestorian.org

Certbot failed to authenticate some domains (authenticator: webroot). The Certificate Authority reported these problems:
  Domain: prototype.genestorian.org
  Type:   unauthorized
  Detail: Invalid response from http://prototype.genestorian.org/.well-known/acme-challenge/xq6moThWqu9cN4loDeBUfVFVcy9A63lX_JTBPL1HgAE [157.245.25.15]: "<!-- No route found for &quot;GET http://prototype.genestorian.org/.well-known/acme-challenge/xq6moThWqu9cN4loDeBUfVFVcy9A63lX_J"

  Domain: www.prototype.genestorian.org
  Type:   unauthorized
  Detail: Invalid response from http://www.prototype.genestorian.org/.well-known/acme-challenge/nG4XwSG9PsCf5wtmD5ta7NCo3QSxHfcMyvyRZ4_7e0s [157.245.25.15]: "<!-- No route found for &quot;GET http://www.prototype.genestorian.org/.well-known/acme-challenge/nG4XwSG9PsCf5wtmD5ta7NCo3QSxHf"

Hint: The Certificate Authority failed to download the temporary challenge files created by Certbot. Ensure that the listed domains serve their content from the provided --webroot-path/-w and that files created there can be downloaded from the internet.

Some challenges have failed.
Ask for help or search for solutions at https://community.letsencrypt.org. See the logfile /var/log/letsencrypt/letsencrypt.log or re-run Certbot with -v for more details.
ERROR: 1
```

I don't think it's a problem with the droplet ports, because I could install certbot in the nginx container and run
```
certbot --nginx -d prototype.genestorian.org -d www.prototype.genestorian.org
```

This worked but then I did not know what else to do, so I figured it was better to find a `docker-compose`  solution.

**Comment by NicolasCARPi** [comment]
Hello Manu.

TLS certs can be a PITA and you will find literally thousands of ways to deal with it with docker. There is no one true way to do it. What I can suggest:

* install certbot on the host, not in the container
* stop the container and use `certbot certonly --standalone -d www.prototype.genestorian.org` (BTW lose the www)
* start the container with a bind mount (volume) on the directory with the cert/privkey so nginx in the container will find it

That will give you 3 months to find a better solution for renewal without having to stop the containers ;)

I like to run HAProxy in front of my containers, and HAProxy takes care of the certbot requests to .well-known, but it's a more involved setup of course.

**Comment by manulera** [comment]
Hello @NicolasCARPi thanks! You are my tooth fairy in this. I post a question at night, and when I come back I find the solution!

I tried this, and installed the certificates, and the nginx server starts, so it can find both the conf and certificates.

However something is not working. When I make an http request with curl to the domain, I get the 301 redirect, which means that the server is reached, I guess. However, when I curl the https `curl https://prototype.genestorian.org/` I get `curl: (7) Failed to connect to prototype.genestorian.org port 443: Connection refused`.

I have allowed https in the firewall 'ufw allow https' on the droplet, and when I do `ss -ntpl` I see that docker is listening in the 443 port.

Something comes to mind? [this](https://github.com/manulera/YeastDatabase/blob/master/phpdocker/nginx/nginx.conf) is the nginx conf.
